### PR TITLE
Normalize error thrown in ObjectQueue `Iterator#next()`

### PR DESCRIPTION
**Pull request by f2prateek**
+ use factory methods instead of implementing classes directly
+ use generics in test

### CODE REVIEW

File: tape/src/main/java/com/squareup/tape/InMemoryObjectQueue.java [hunk 1]
```diff
@@ -102,6 +102,8 @@
       checkForComodification();
 
       if (size() == 0) throw new NoSuchElementException();
+      if (nextElementIndex >= size()) throw new NoSuchElementException();
```
Comment (swankjesse) [inline tape/src/main/java/com/squareup/tape/InMemoryObjectQueue.java]: Does this make the check on the previous line unnecessary?

**Comment by swankjesse** [comment]
LGTM

**Comment by JakeWharton** [comment]
LGTM, too

File: tape/src/main/java/com/squareup/tape/InMemoryObjectQueue.java (hunk 1 above)
Comment (f2prateek) [inline tape/src/main/java/com/squareup/tape/InMemoryObjectQueue.java]: You're right!

**Pull request merged**